<launch>
  <!-- shelf-->
  <arg name="json" default="$(find jsk_2015_05_baxter_apc)/json/layout_12.json"/>
  <arg name="UPPER_SHELF" value="$(find jsk_apc2016_common)/config/bin_upper_shelf.yaml"/>
  <arg name="LOWER_SHELF" value="$(find jsk_apc2016_common)/config/bin_lower_shelf.yaml"/>

  <!-- camera-->
  <arg name="INPUT_CLOUD" value="/kinect2_torso/hd/points"/>
  <arg name="INPUT_INFO" value="/kinect2_torso/hd/camera_info"/>
  <arg name="INPUT_IMAGE" value="/kinect2_torso/hd/image_color"/>

  <!-- target bin-->
  <arg name="INPUT_BIN_ARRAY" value="/set_bin_param/bin_array"/>
  <arg name="INPUT_TARGET" value="/right_hand/target_bin"/>
  <arg name="TARGET_BIN" value="g"/>
  <arg name="TRAINED_PATH" value="$(find jsk_apc2016_common)/data/trained_segmenter.pkl" />

  <param name="/right_hand/target_bin" value="$(arg TARGET_BIN)"/>

  <node pkg="jsk_apc2016_common" type="publish_bin_info.py"
    name="set_bin_param">
    <param name="json" value="$(arg json)"/>
    <rosparam file="$(arg UPPER_SHELF)" command="load" ns="upper_shelf"/>
    <rosparam file="$(arg LOWER_SHELF)" command="load" ns="lower_shelf"/>
  </node>

  <node pkg="jsk_apc2016_common" type="publish_target_bin_info.py"
    name="pub_target_bin_info" output="screen">
    <remap from="~target_bin_name" to="$(arg INPUT_TARGET)"/>
    <remap from="~input/bin_info_array" to="$(arg INPUT_BIN_ARRAY)"/>
  </node>

  <node pkg="jsk_apc2016_common" type="publish_bin_tf.py"
    name="pub_bin_tf">
    <remap from="~input/bin_info_array" to="$(arg INPUT_BIN_ARRAY)"/>
  </node>

  <node pkg="jsk_apc2016_common" type="cloud_to_spatial_features"
    name="sib_spatial">
    <remap from="~input" to="$(arg INPUT_CLOUD)"/>
    <remap from="~target_bin" to="/pub_target_bin_info/target_bin_info"/>
  </node>

  <node pkg="jsk_apc2016_common" type="tf_bbox_to_mask.py"
      name="sib_mask" output="screen">
    <remap from="~input/bin_info_array" to="$(arg INPUT_BIN_ARRAY)"/>
    <remap from="~target_bin_name" to="$(arg INPUT_TARGET)"/>
    <remap from="~input" to="$(arg INPUT_INFO)"/>
  </node>

  <node pkg="jsk_apc2016_common" type="sib_topic_synchronizer"
        name="sib_topic_sync" output="screen">
    <remap from="~input/dist" to="/sib_spatial/dist"/>
    <remap from="~input/height" to="/sib_spatial/height"/>
    <remap from="~input/image" to="$(arg INPUT_IMAGE)"/>
    <remap from="~input/mask" to="/sib_mask/output"/>
  </node>

  <node pkg="jsk_apc2016_common" type="rbo_segmentation_in_bin_node.py"
        name="sib_rbo" output="screen">
    <remap from="~input" to="sib_topic_sync/output"/>
    <remap from="~input/bin_info_array" to="$(arg INPUT_BIN_ARRAY)"/>
    <remap from="~target_bin_name" to="$(arg INPUT_TARGET)"/>
    <param name="trained_pkl_path" value="$(arg TRAINED_PATH)" />
    <param name="always_subscribe" value="True"/>
  </node>

  <arg name="MANAGER" value="segmentation_in_bin_manager" />
  <node name="$(arg MANAGER)"
        pkg="nodelet" type="nodelet" args="manager" />

  <node name="sib_mask_image_to_pi"
        pkg="nodelet" type="nodelet"
        args="load jsk_pcl/MaskImageToPointIndices $(arg MANAGER)">
    <remap from="~input" to="sib_rbo/target_mask"/>
  </node>

  <node name="sib_extract_pi"
        pkg="nodelet" type="nodelet"
        args="load jsk_pcl/ExtractIndices $(arg MANAGER)">
    <remap from="~input" to="$(arg INPUT_CLOUD)" />
    <remap from="~indices" to="sib_mask_image_to_pi/output" />
    <rosparam>
      max_queue_size: 100
      approximate_sync: true
    </rosparam>
  </node>

  <node name="sib_centroid_publisher"
        pkg="nodelet" type="nodelet"
        args="load jsk_pcl/CentroidPublisher $(arg MANAGER)">
        <remap from="~input" to="sib_extract_pi/output"/>
  </node>

  <include file="$(find jsk_2016_01_baxter_apc)/launch/include/sib_visualization.launch">
      <arg name="POSTERIOR" value="/sib_rbo/posterior"/>
      <arg name="INPUT_IMAGE" value="/kinect2_torso/hd/image_color"/>
      <arg name="MASK" value="/sib_rbo/target_mask"/>
      <arg name="HAND" value="torso"/>
  </include>
</launch>
